### PR TITLE
Fix client side tag violations and Improve function signatures

### PR DESCRIPTION
**Pull request by cead22**
### Details

- This started as a follow up to this PR https://github.com/Expensify/App/pull/33969 to address comments we didn't get to address before the end of last week when Infinite Red finished working with us on violations. Namely, updating the method signatures of TransactionUtils.hasViolation, TransactionUtils.getTransactionViolations, and ReportUtils.isSettled
- I was getting `Malformed currency code` when testing ([see slack thread in expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q/p1708384127377139)), so I added a fix to this PR
- I also noticed the missing tag violation wasn't working and dug into it, and added a fix as well (I believe it was broken here https://github.com/Expensify/App/pull/34983)

### Fixed Issues

$ https://github.com/Expensify/App/issues/36889
- Followup to https://github.com/Expensify/App/pull/33969
- PROPOSAL: N/A

### Tests
**Setup**
- Make sure emails sent to the testing email addresses go to an inbox you can access to validate your account(s)
- Add the logins you’re using to test to the violation beta
- Create a new collect policy on OldDot and; 
    - Set the approval mode to “submit and approve”
    - Invite an additional user to the policy to play the member role in these testing flows
    - From the policy editor, run the script below from the workspace editor in old dot to enable the policy for workspace creation for a paid plan.
    - From the policy editor > Categories, switch on “People must categorize expenses”, and add a some categories
    - From the policy editor > Tags, switch on “People must tag expenses”, and add some tags

```js
p = Policy.getCurrent();
p.policy.isPolicyExpenseChatEnabled = "true";
p.save();
```

**Test**
- Tap big plus sign > request money
- Select Manual tab (from Manual | Scan | Distance)
- Enter amount and tap next
- Tap the workspace from the pre-testing steps (tap the row, not the split button in the row)
- Tap the show more button if it’s there
    - [x] Verify there’s a Required label on the Category row
    - [x] Verify there’s a Required label on the tag row
- Tap the Request $X.XX button
    - [x] Verify the report preview in the workspace chat has an RBR
- Tap the report preview
    - [x] Verify the money request preview in the iou report has an RBR
- Tap the money request preview
    - [x] Verify you see the Missing Category red text under the category row
- Tap the category row
- Tap on a category
    - [x] Verify you no longer see the Missing Category red text under the category row
- Tap the Tag row
- Tap on a tag
    - [x] Verify you no longer see the Missing Tag red text under the category row
- Tap on the IOU chat report link at the top
    - [x] Verify the money request preview in the iou report no longer has an RBR
- Tap on the policy expense chat report link at the top
    - [x] Verify the report preview in the policy expense chat report no longer has an RBR

---

- [x] Verify that no errors appear in the JS console

### Offline tests
Same as test

### QA Steps
Same as tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/165133/7cdac983-ecf8-4f80-a1ca-45b43ffce74c

</details>

<details>
<summary>Android: mWeb Chrome</summary>

**Note** I'm still trying to figure out how to test this, and will rely on the reviewer checklist for this
</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/165133/0664f4a8-ef2b-4edf-bdfc-648adeb73687

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/165133/a23bca35-142e-40e4-8fe8-6dc286e4db1f

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/165133/4529f6f4-8d21-48c8-b1b6-ff30e3a2773e

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/165133/d56ff54c-4fec-409b-94b4-76632cbfaf2a

</details>

### CODE REVIEW

**Comment by allroundexperts** [comment]
@cead22 Am I supposed to review this? This one seems to be a WIP.

**Comment by cead22** [comment]
@allroundexperts no, or not yet at least, I assigned you and Yuwen as reviewers cause you had context on the deploy blocker. I could've just cc'ed you, but if you want to be the C+ to review this PR, that works

**Comment by s77rt** [comment]
@cead22 

> Tap the Request $X.XX button
> Verify the report preview in the workspace chat has an RBR

We shouldn't let the user proceed with creating the request if the required fields are missing

**Comment by cead22** [comment]
> We shouldn't let the user proceed with creating the request if the required fields are missing

No that's not true, users should be able to create money requests without entering category and tag even if they're required. And for merchant, we already prevent it

**Comment by s77rt** [comment]
@cead22 So they are required in the term that they cannot be submitted?

**Comment by cead22** [comment]
They can also be submitted, approved, and paid. They're used to highlight violations to users submitting the requests, and those approving them as well

**Comment by s77rt** [comment]
Choosing a tag for multi-level tags does not clear the error

https://github.com/Expensify/App/assets/16493223/ea2eff27-0beb-4de4-be43-e2dcff2760e2

**Comment by s77rt** [comment]
Setting a category value clears the error right away, while the tag error takes a while (waits for BE)



https://github.com/Expensify/App/assets/16493223/359f5a12-df96-44fa-afbc-44ef77b88919

**Comment by s77rt** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If the PR modifies the form input styles:
    - [X] I verified that all the inputs inside a form are aligned with each other.
    - [X] I added `Design` label so the design team can review the changes.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



https://github.com/Expensify/App/assets/16493223/0cdad8ed-d89a-43f1-8b97-1f13f5a6ff89


</details>

<details>
<summary>Android: mWeb Chrome</summary>



https://github.com/Expensify/App/assets/16493223/bca748d8-6236-4eca-a559-19021836b3b9


</details>

<details>
<summary>iOS: Native</summary>



https://github.com/Expensify/App/assets/16493223/68a65290-e558-40b1-8aa5-817361d58930


</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/Expensify/App/assets/16493223/3bb40c04-16b1-4777-9778-5a844015716a


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/Expensify/App/assets/16493223/4e694c16-57dc-46ac-b270-c3d037accbf7


</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/Expensify/App/assets/16493223/499bcb48-4a66-4959-81a6-80c56fc5ec27


</details>

**Comment by cead22** [comment]
@rlinoz @s77rt please hold on reviewing. Fixing the tag violations for the multi-tag case is trickier than I thought

**Comment by cead22** [comment]
@rlinoz @s77rt this is ready for review. Please note that I remove code for violations for multi-level tags to avoid bloating the scope of this PR, and I'll follow up with another PR to add violation support for multi tags

**Comment by rlinoz** [comment]
CategoryPicker is broken for me
`Uncaught TypeError: Cannot read properties of undefined (reading 'enabled')`, but it is also happening on main.

I will try to understand what is happening tomorrow.

**Comment by cead22** [comment]
@rlinoz I'm pretty sure this is known and happening for the tags as well, which is being fixed here https://github.com/Expensify/App/pull/36994/files

**Comment by s77rt** [comment]
For multi level tags still seeing RBR in the money request preview - Do you want to handle this in the follow up as well?


https://github.com/Expensify/App/assets/16493223/decb2260-4ff3-43b2-999f-df9d81636fc1

**Comment by s77rt** [comment]
For single tag, the tag selection works as expected but cannot test category due to other bugs

**Comment by s77rt** [comment]
@cead22 Can you please merge main and resolve the conflicts?

**Comment by cead22** [comment]
@s77rt merged main

> For multi level tags still seeing RBR in the money request preview - Do you want to handle this in the follow up as well?

Yes

File: src/libs/Violations/ViolationsUtils.ts [hunk 1]
```diff
@@ -46,72 +45,40 @@ const ViolationsUtils = {
...
-                const hasTagOutOfPolicyViolation = transactionViolations.some((violation) => violation.name === CONST.VIOLATIONS.TAG_OUT_OF_POLICY && violation.data?.tagName === key);
-                const hasMissingTagViolation = transactionViolations.some((violation) => violation.name === CONST.VIOLATIONS.MISSING_TAG && violation.data?.tagName === key);
-                const selectedTag = selectedTags[index];
-                const isTagInPolicy = Boolean(policyTagList[key]?.tags[selectedTag]?.enabled);
+            // At the moment, we only return violations for tags for workspaces with single-level tags
+            if (policyTagKeys.length === 1) {
+                const policyTagListName = Object.keys(policyTagList)[0];
+                const policyTags = policyTagList[policyTagListName]?.tags;
+                const hasTagOutOfPolicyViolation = transactionViolations.some((violation) => violation.name === 'tagOutOfPolicy');
+                const hasMissingTagViolation = transactionViolations.some((violation) => violation.name === 'missingTag');
```
Comment (s77rt) [inline src/libs/Violations/ViolationsUtils.ts:60]:
```suggestion
                const hasTagOutOfPolicyViolation = transactionViolations.some((violation) => violation.name === CONST.VIOLATIONS.TAG_OUT_OF_POLICY);
                const hasMissingTagViolation = transactionViolations.some((violation) => violation.name === CONST.VIOLATIONS.MISSING_TAG);
```

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/hooks/useViolations.ts [hunk 2]
```diff
@@ -58,19 +58,7 @@ function useViolations(violations: TransactionViolation[]) {
...
         return violationGroups ?? new Map();
     }, [violations]);
-
-    const getViolationsForField = useCallback(
-        (field: ViolationField, data?: TransactionViolation['data']) => {
-            const currentViolations = violationsByField.get(field) ?? [];
-
-            if (data?.tagName) {
-                return currentViolations.filter((violation) => violation.data?.tagName === data.tagName);
-            }
```
[the remark this answers is not in the document]
Comment (cead22) [reply src/hooks/useViolations.ts:68]: I created an issue to follow up to implement violations for multi level tags https://github.com/Expensify/App/issues/37117

**Comment by cead22** [comment]
@s77rt I committed your suggestions

**Comment by cead22** [comment]
Yeah that change was intentional. I created an issue to follow up to implement violations for multi level tags https://github.com/Expensify/App/issues/37117

**Comment by s77rt** [comment]
That removed condition is for empty tags list, does not seem related to the multi tags.

**Comment by cead22** [comment]
Is there a specific case you're worried about not working? I tested again and things seem to work fine. There's a couple of non-obvious things at play here
1. The removed condition is checking for empty tag lists on a workspace, and I don't think that can ever happen. You may have a workspace with a single tag list with no tags enabled, but there will still be a tag list
2. Even if the workspace had no tag lists, the workspace's "requires tag" property/logic should return false. Similarly for the case of a single tag lists that has tags in it, but where all the tags are disabled, the workspace "requires tag" property/logic should return false

So for the case where there is only 1 tag list, with no tags enabled, and the money request has a tag selected (that is disabled), we don't go into the `if (policyRequiresTags) {`

I say `"requires tag" property/logic should return false` because that may not be the case on the new dot client, but it is the case in our php logic, which looks like the code below

```php
    public function requiresTag()
    {
...
        if (!$this->hasEnabledTags()) {
            return false;
        }
...
    public function hasEnabledTags()
    {
...
        // If the policy has independent tags, ensure we check each tag list to see if there are any required tags
        if ($this->hasIndependentTags()) {
            return ArrayUtils::some($this->getTagListsArray(), function (array $tagList, int $i): bool {
                $tagCollection = $this->getIndependentTags($i);
                return $tagCollection->hasEnabledTags();
            });
        }
```

**Review by s77rt — APPROVED** [review]

**Comment by cead22** [comment]
@neil-marcellini not sure why you were added here, but a review from either you or @rlinoz works for me :)

**Review by rlinoz — APPROVED** [review]
Code looks good and tests well, thanks!

**Comment by cead22** [comment]
Thank you! Scratch my previous request @neil-marcellini

**Pull request merged**